src/goto-instrument/function_stubs.cpp

```diff
@@ -0,0 +1,264 @@
+    goto_function.body.add(goto_programt::make_decl(r));
+    call.lhs() = r;
+
+    // Replace the return_value with r
+    symbol_exprt ret_val(CPROVER_PREFIX "return_value", type.return_type());
+    replace.insert(ret_val, r);
+  }
+
+  // Add the parameters as arguments to the call
+  auto parameter_identifiers = type.parameter_identifiers();
+  for(auto it = parameter_identifiers.begin();
+      it != parameter_identifiers.end();
+      it++)
+  {
+    const symbolt &parameter_symbol = ns.lookup(*it);
+    call.arguments().push_back(parameter_symbol.symbol_expr());
+  }
+
+  goto_function.body.add(
+    goto_programt::make_function_call(code_function_callt(call)));
+  // Maybe: Add location/comments maybe?
+
+  // Add the postcondition assumption
+  if(ensures.is_not_nil())
+  {
+    exprt ensures_replaced = ensures;
+    replace(ensures_replaced);
+    goto_function.body.add(goto_programt::make_assumption(
+      ensures_replaced, ensures.source_location()));
+  }
+
+  if(type.return_type() != empty_typet())
+  {
+    code_returnt return_expr(r);
+    goto_function.body.add(goto_programt::make_return(return_expr));
+  }
+
+  // Append the end of function instruction
+  goto_function.body.add(
+    goto_programt::make_end_function(ensures.source_location()));
+
+  log.debug() << "Function body after adding the assumption: "
+              << "\n";
+  goto_function.body.output(ns, function_id, log.debug());
+
+  // Question: do I need to call some update or something after I do
+  // the changes?
+}
+
+bool function_stubst::operator()()
+{
+  // Get the body of the requested function to stub
+  goto_functiont function_body;
+  auto requested_function_body =
+    goto_functions.function_map.find(function_name);
+  if(requested_function_body == goto_functions.function_map.end())
+  {
+    log.error() << "The body of function: " << function_name
+                << "doesn't exist in the goto_model\n";
+    return true;
+  }
+  function_body.copy_from(requested_function_body->second);
+
+  log.debug() << "Function: " << function_name << "\n";
+
+  // Add a declaration of the impl function, that will be called in
+  // the function body. Create a new function symbol for the internal
+  // implementation function
+  //
+  // Question: Is it enough to add it in the symbol_table?
+  //
+  // Question: Should I make lookups on the symbol table or should I
+  // make them on the namespace?
+  if(!symbol_table.has_symbol(function_name))
+  {
+    log.error() << "Function: " << function_name
+                << "doesn't exist in the symbol table\n";
+    return true;
+  }
+  symbolt function_symbol = symbol_table.lookup_ref(function_name);
+  const irep_idt stub_function_name = stub_name_of_function(function_name);
+  function_symbol.name = stub_function_name;
+  function_symbol.location = source_locationt();
+
+  // Question: Maybe I should also change the module?
+
+  // Question: Is it correct to change all names with the same name?
```

Comment:
It's harmless AFAICT. Pretty output might be a little less pretty than is ideal.